cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+    """
+    Schedules the tasks of bears to the process pool and run the event loop to
+    complete those tasks.
+    :param bears:          A list of bears to be scheduled onto the process
+                           pool.
+    :param dep_resolver:   The object that keeps track of dependencies.
+    :param event_loop:     The asyncio event loop.
+    :param blocking_tasks: Dictionary that keeps track of the remaining tasks
+                           of each bear.
+    :param executor:       The process pool to which the bear tasks are
+                           scheduled.
+    """
+    for bear in bears:
+        if bear not in dep_resolver.dependency_dict:
+            blocking_tasks[bear] = [
+                event_loop.run_in_executor(executor, bear.execute_task, task)
+                for task in bear.generate_tasks()]
+            for task in blocking_tasks[bear]:
+                task.add_done_callback(functools.partial(finish_task,
+                                                         bear,
+                                                         dep_resolver,
+                                                         blocking_tasks,
+                                                         event_loop,
+                                                         executor))
+
+
+def finish_task(bear,
+                dep_resolver,
+                blocking_tasks,
+                event_loop,
+                executor,
+                task):
+    """
+    The callback for when a task of a bear completes. It is responsible for
+    checking if the bear completed it's execution and the handling of the
+    result generated by the task.
+    :param bear:           The bear that the task belongs to.
+    :param dep_resolver:   The object that keeps track of dependencies.
+    :param blocking_tasks: Dictionary that keeps track of the remaining tasks
+                           of each bear.
+    :param event_loop:     The asyncio event loop.
+    :param executor:       The process pool to which the bear tasks are
+                           scheduled.
+    :param task:           The task that completed.
+    """
+    log = logging.getLogger('Consumer task')
+    log.info(bear)
+    blocking_tasks[bear].remove(task)
+    if not blocking_tasks[bear]:
+        resolved_bears = dep_resolver.resolve(bear)
+        if resolved_bears:
+            schedule_bears(resolved_bears,
+                           dep_resolver,
+                           event_loop,
+                           blocking_tasks,
+                           executor)
+        del blocking_tasks[bear]
+    if not blocking_tasks:
+        event_loop.stop()
+
+
+def main():
+    # Configure logging to show the name of the thread where the log
+    # message originates.
+    logging.basicConfig(
+        level=logging.INFO,
+        format='%(process)5s %(name)18s: %(message)s',
+        stream=sys.stderr,
+    )
+
+    dep_resolver = Dependencies()
```

Comment:
some lines below you use the variable first, no need to instantiate the class, do something else and then use it ;D